File: general-server/src/main/java/com/song/middleware/server/storage/mysql/MysqlStorageComponent.java
package com.song.middleware.server.storage.mysql;

import com.song.middleware.server.domain.ServiceDO;
import com.song.middleware.server.domain.ServiceStatus;
import com.song.middleware.server.storage.ServiceStorageComponent;
import com.song.middleware.server.storage.exception.StorageException;
import com.song.middleware.server.storage.mysql.mapper.ServiceMapper;
import com.song.middleware.server.util.HttpUtils;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import javax.annotation.PostConstruct;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Request.Builder;
import okhttp3.Response;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

/**
 * Created by song on 2017/8/5.
 */
@Slf4j
@Component
public class MysqlStorageComponent implements ServiceStorageComponent {

    private static final int DEFAULT_CAPACITY = 5000;

    private final ServiceMapper serviceMapper;

    private ExecutorService executor;

    private ScheduledExecutorService healthCheckExecutor;

    @Autowired
    public MysqlStorageComponent(ServiceMapper serviceMapper) {
        this.serviceMapper = serviceMapper;
    }

    @PostConstruct
    public void init() {
        int processorNum = Runtime.getRuntime().availableProcessors();
        this.executor = new ThreadPoolExecutor(processorNum, processorNum, 0L,
            TimeUnit.MILLISECONDS,
            new LinkedBlockingQueue<>(DEFAULT_CAPACITY), (r, executor) -> {
            executor.submit(r);
            throw new StorageException("The storage thread pool queue is full, plz check.");
        });
        this.healthCheckExecutor = Executors.newScheduledThreadPool(processorNum);
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            executor.shutdown();
            try {
                if (!executor.awaitTermination(1, TimeUnit.MINUTES)) {
                    List<Runnable> runnables = executor.shutdownNow();
                    log.warn(
                        "Storage executor was abruptly shut down. {} running tasks will be interrupted.",
                        runnables.size());
                }
                healthCheckExecutor.shutdown();
            } catch (InterruptedException ignore) {
                //we're shutting down anyway.
            }
        }));
    }

    @Override
    public ServiceDO syncQuery(String name, String ip, String port) {
        try {
            return serviceMapper.findOne(name, ip, port);
        } catch (StorageException e) {
            throw new StorageException(e);
        }
    }

    @Override
    public List<ServiceDO> syncQuery(String name) {
        try {
            return serviceMapper.findAllByName(name);
        } catch (Exception e) {
            throw new StorageException(e);
        }
    }

    @Override
    public Future<List<ServiceDO>> asyncQuery(final String name) {
        return executor.submit(() -> serviceMapper.findAllByName(name));
    }

    @Override
    public void syncRegister(ServiceDO serviceDO) {
        try {
            ServiceDO serviceInDB = serviceMapper
                .findOne(serviceDO.getName(), serviceDO.getIp(), serviceDO.getPort());
            if (serviceInDB == null) {
                /*
                    Even if there is an race condition, mysql will complain about it.
                 */
                serviceDO.setStatus(ServiceStatus.UP.getCode());
                serviceMapper.save(serviceDO);
            } else if (serviceDO.getStatus() == ServiceStatus.DOWN.getCode()) {
                serviceMapper
                    .updateStatus(serviceInDB.getId(), ServiceStatus.DOWN.getCode());
            }
            this.healthCheckExecutor.scheduleAtFixedRate(() -> {
                ServiceDO serviceInstance = serviceMapper
                    .findOne(serviceDO.getName(), serviceDO.getIp(), serviceDO.getPort());
                boolean checkSuccess = false;
                try {
                    Response execute = HttpUtils.getInstance()
                        .newCall(
                            new Builder().url(serviceInstance.getHealthCheckUrl()).build())
                        .execute();
                    if (execute.isSuccessful()) {
                        checkSuccess = true;
                    }
                } catch (IOException e) {
                    log.error("health check failed", e);
                }
                if (!checkSuccess) {
                    serviceMapper
                        .updateStatus(serviceInstance.getId(), ServiceStatus.DOWN.getCode());
                }
            }, 60, serviceDO.getHealthCheckInterval(), TimeUnit.SECONDS);
            log.info("Discover a new instance {}/{}:{}", serviceDO.getName(), serviceDO.getId(),
                serviceDO.getIp());
        } catch (Exception e) {
            throw new StorageException("Save service into storage failed_" + e.getMessage(), e);
        }
    }

    @Override
    public Future<?> asyncRegister(ServiceDO serviceDO) {
        return executor.submit(() -> syncRegister(serviceDO));
    }

    @Override
    public void syncUnregister(String name, String ip, String port) {
        try {
            ServiceDO serviceDO = serviceMapper.findOne(name, ip, port);
            if (serviceDO != null) {
                serviceMapper.updateStatus(serviceDO.getId(), ServiceStatus.OFFLINE.getCode());
            }
        }catch (Exception e){
            throw new StorageException(e);
        }
    }

}

File: general-server/src/main/java/com/song/middleware/server/storage/mysql/MysqlStorageComponent.kt
package com.song.middleware.server.storage.mysql

import com.song.middleware.server.domain.ServiceDO
import com.song.middleware.server.domain.ServiceStatus
import com.song.middleware.server.storage.ServiceStorageComponent
import com.song.middleware.server.storage.exception.StorageException
import com.song.middleware.server.storage.mysql.mapper.ServiceMapper
import com.song.middleware.server.util.HttpUtils
import okhttp3.Request.Builder
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.io.IOException
import java.util.concurrent.*
import javax.annotation.PostConstruct

/**
 * Created by song on 2017/8/5.
 */
@Component
class MysqlStorageComponent @Autowired
constructor(private val serviceMapper: ServiceMapper) : ServiceStorageComponent {

    private var executor: ExecutorService? = null

    private var healthCheckExecutor: ScheduledExecutorService? = null

    @PostConstruct
    fun init() {
        val processorNum = Runtime.getRuntime().availableProcessors()
        this.executor = ThreadPoolExecutor(processorNum, processorNum, 0L,
                TimeUnit.MILLISECONDS,
                LinkedBlockingQueue<Runnable>(DEFAULT_CAPACITY)) { r, executor ->
            executor.submit(r)
            throw StorageException("The storage thread pool queue is full, plz check.")
        }
        this.healthCheckExecutor = Executors.newScheduledThreadPool(processorNum)
        Runtime.getRuntime().addShutdownHook(Thread {
            executor!!.shutdown()
            try {
                if (!executor!!.awaitTermination(1, TimeUnit.MINUTES)) {
                    val runnables = executor!!.shutdownNow()
                    log.warn(
                            "Storage executor was abruptly shut down. {} running tasks will be interrupted.",
                            runnables.size)
                }
                healthCheckExecutor!!.shutdown()
            } catch (ignore: InterruptedException) {
                //we're shutting down anyway.
            }
        })
    }

    override fun syncQuery(name: String, ip: String, port: String): ServiceDO {
        try {
            return serviceMapper.findOne(name, ip, port)
        } catch (e: StorageException) {
            throw StorageException(e)
        }

    }

    override fun syncQuery(name: String): List<ServiceDO> {
        try {
            return serviceMapper.findAllByName(name)
        } catch (e: Exception) {
            throw StorageException(e)
        }

    }

    override fun asyncQuery(name: String): Future<List<ServiceDO>> {
        return executor!!.submit<List<ServiceDO>> { serviceMapper.findAllByName(name) }
    }

    override fun syncRegister(serviceDO: ServiceDO) {
        try {
            val serviceInDB = serviceMapper
                    .findOne(serviceDO.name!!, serviceDO.ip!!, serviceDO.port!!)
            if (serviceDO.status == ServiceStatus.DOWN.code) {
                serviceMapper
                        .updateStatus(serviceInDB.id, ServiceStatus.DOWN.code)
            }
            this.healthCheckExecutor!!.scheduleAtFixedRate({
                val serviceInstance = serviceMapper
                        .findOne(serviceDO.name!!, serviceDO.ip!!, serviceDO.port!!)
                var checkSuccess = false
                try {
                    val execute = HttpUtils.instance
                            .newCall(
                                    Builder().url(serviceInstance.healthCheckUrl).build())
                            .execute()
                    if (execute.isSuccessful) {
                        checkSuccess = true
                    }
                } catch (e: IOException) {
                    log.error("health check failed", e)
                }

                if (!checkSuccess) {
                    serviceMapper
                            .updateStatus(serviceInstance.id, ServiceStatus.DOWN.code)
                }
            }, 60, serviceDO.healthCheckInterval, TimeUnit.SECONDS)
            log.info("Discover a new instance {}/{}:{}", serviceDO.name, serviceDO.id,
                    serviceDO.ip)
        } catch (e: Exception) {
            throw StorageException("Save service into storage failed_" + e.message, e)
        }

    }

    override fun asyncRegister(serviceDO: ServiceDO): Future<*> {
        return executor!!.submit { syncRegister(serviceDO) }
    }

    override fun syncUnregister(name: String, ip: String, port: String) {
        try {
            val serviceDO = serviceMapper.findOne(name, ip, port)
            if (serviceDO != null) {
                serviceMapper.updateStatus(serviceDO.id, ServiceStatus.OFFLINE.code)
            }
        } catch (e: Exception) {
            throw StorageException(e)
        }

    }

    companion object {

        private val log = LoggerFactory.getLogger(MysqlStorageComponent.javaClass)

        private val DEFAULT_CAPACITY = 5000
    }

}
